keps/sig-etcd/4578-feature-gate/README.md

```diff
@@ -0,0 +1,388 @@
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+New feature gate unit tests will be added.
+
+##### Integration tests
+
+New feature gate integration tests will be added.
+
+##### e2e tests
+
+We will add one cluster level and one server level experimental features to the feature gate (without removing the original experimental flag), and add e2e tests to make sure the feature gate functions equivalently to their `--experimental-xxx` flags.
+
+We will also add downgrade/upgrade e2e tests to make sure the feature gate does not break the cluster in mixed version cluster.
+
+### Graduation Criteria
+
+#### Milestone 1
+
+* server level feature gate implemented.
+* cluster level feature gate implemented.
+
+#### Milestone 2
+
+* backport necessary proto changes to 3.5.
+* http endpoints added.
+* grpc endpoints added.
+* `etcdctl` command added.
+* equivalent of one cluster level and one server level experimental features added to the feature gate.
+* initial e2e tests completed.
+
+#### Milestone 3
+
+* e2e tests for upgrade & downgrade & mixed version scenarios completed.
+* robustness tests completed. 
+* documentation.
+
+### Upgrade / Downgrade Strategy
+
+The feature gate feature would available in 3.6+. 
+In 3.5, the necessary proto changes would be backported, but feature gate would not be available. 
+When upgrading/downgrading between 3.5 and 3.6, no feature would be enabled if there is at least one member with version 3.5.
+
+If a user uses a 3.5 version that does not recognize the new proto field, it is recommended they should upgrade to the latest 3.5 release first before upgrading to 3.6. Since unknown fields is ignored by proto, etcd should still work even if they do not upgrade to latest 3.5 version.
+
+### Version Skew Strategy
+
+Since we are introducing new fields in `member.Attributes` which is sent through raft, we will backport the new fields into 3.5, but would not use them in any way affecting 3.5 etcd server.
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
```

Comment:
We could introduce the internal feature gate support in a minor version of etcd (eg 3.6), so that an 3.6 etcd can see what feature gates are active. And then release etcd 3.7 that actually lets you turn on feature gates.

I think that would be more effort but a better downgrade story.

Reply:
etcd has very long release cycles. If possible, it would be good to have the feature gate available in 3.6. That's why we are proposing to have minimal backport in 3.5, so that it is ok to start using them in 3.6.